### PR TITLE
[mesheryctl] Feature: Added helper function for current-context support 

### LINKED ISSUE #2393: [mesheryctl] Respect current-context, fetch correct docker-compose.yaml
#### Description

Whenever `mesheryctl` cannot find `docker-compose.yaml` it downloads it from Github, but all it does not take into account the `correct-context` and always download the latest file from the master branch.
[Several instances](https://github.com/layer5io/meshery/search?q=fileURL) of this behaviour needs to be corrected.

`mesheryctl system reset` has been updated to fetch correct `docker-compose.yaml`  file from GitHub corresponding to the `channel` and `version` specified in the `current-context`. 

The logic inside `resetMesheryConfig` can be extracted to create a separate helper function that fetches correct file.
https://github.com/layer5io/meshery/blob/1a91051afe17d82a668c7429c735bf7ed94970c7/mesheryctl/internal/cli/root/system/reset.go#L80-L90

 

#### Expected Behavior

`mesheryctl` should fetch correct `docker-compose.yaml` based on `current-context`

#### Child Issues

Issue #2394 [mesheryctl] Feature: Add current-context support to `system update`
Issue #2395 [mesheryctl] Feature: Add current-context support to `system log`
Issye #2396 [mesheryctl] Feature: Add current-context support to `system start`

### LINKED ISSUE #2394: [mesheryctl] Feature: Add current-context support to `system update`
#### Description

See Epic #2393 for more context

#### Current Behavior

https://github.com/layer5io/meshery/blob/1a91051afe17d82a668c7429c735bf7ed94970c7/mesheryctl/internal/cli/root/system/update.go#L44-L48

#### Expected Behavior

`mesheryctl system update` should fetch correct `docker-compose.yaml` as per `version` and `channel` specified in `current-context`

### PR DESCRIPTION
**Pull request by aneeshnema**
**Description**

Added helper function to resolve child issues of #2393 
Refactored `resetMesheryConfig` to use the new helper functions `DownloadDockerComposeFile` and `GetCurrentContext`

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Review by navendu-pottekkat — COMMENTED** [review]
Shouldn't we also consider for the temporary changes in the context? This is implemented in the `system reset` command right now and will also be added to the other commands. 

```
currentContext := mctlCfg.CurrentContext

		// if the user has mentioned a temporary context in the -c flag, change the context and proceed to reset
		if tempContext != "" {
			if _, val := mctlCfg.Contexts[tempContext]; val {
				currentContext = tempContext
			} else {
				// if the user specifies a context that is not in the config.yaml file, throw an error and show the available contexts
				log.Errorf("\n\"%s\" context does not exist. The available contexts are:", tempContext)
				for context := range mctlCfg.Contexts {
					log.Errorf("%s", context)
				}
				return nil
			}
		}
```

One suggestion for this would be to take in the current context as an argument in the function.

**Comment by navendu-pottekkat** [comment]
@aneeshnema merging this PR may close the epic issue since you have mentioned "this PR fixes"...

File: mesheryctl/internal/cli/root/system/reset.go
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/reset.go:60]: Why do we need to pass in the tempContext? We are already getting the current context in line 47. Once we have the current-context or temp context, we just need to pass that to the `DownloadDockerComposeFile()`function.

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [inline mesheryctl/pkg/utils/helpers.go:724]: Why do we need to make this call again since we have already picked the current context or the temporary context before?

**Review by navendu-pottekkat — COMMENTED** [review]
@aneeshnema This may require some changes. I have made some comments on the code and on the first run, I got an error where in my `config.yaml` file, I have not specified the version and I am getting an output like this-

Screenshot from 2021-02-11 17-25-19

Could you please check these?

File: mesheryctl/internal/cli/root/system/reset.go
Comment (aneeshnema) [reply mesheryctl/internal/cli/root/system/reset.go:60]: The current context in line 47 was just for getting the name, channel and version of the context. These values are later logged in the function see line 53-56.

File: mesheryctl/pkg/utils/helpers.go
Comment (aneeshnema) [reply mesheryctl/pkg/utils/helpers.go:724]: Yes we did already get the context, but that is only in the case of `resetMesheryConfig`, it was done for logging purpose as mentioned in the other comment. For the rest of the areas where `DownloadDockerComposeFile` will be used, we will not find the `current-context` beforehand.

**Comment by aneeshnema** [comment]
@navendu-pottekkat its' currently logging the values specified current-context, previously if the version was unspecified the latest version would be downloaded. Now after the refactoring, that logic has been moved to `DownloadDockerComposeFile`. Now to solve this problem, we have two ways:

1. Add that logic back to `resetMesheryConfig` additionally(redundant).
2. Add the logging to `DownloadDockerComposeFile`, but this will log anytime if the docker-compose.yaml file will be downloaded (for all commands)

I personally prefer the 2nd option as it would be good to show the user the configuration they are using, and this would only be one time (since you wouldn't download the file again).

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [reply mesheryctl/pkg/utils/helpers.go:724]: In those cases too, can we use the same logic as we used here? One function which takes in the temporary context and returns either the temporary context or the current context. Another function which takes in the current/temporary context (whichever is set) and then downloads the corresponding version file. I think that approach would be neat and readable.

Adding @anirudhjain75 to this thread.

**Comment by navendu-pottekkat** [comment]
@aneeshnema I think we need the log only for the reset command, and maybe the function can return the version if the version is unspecified.

File: mesheryctl/pkg/utils/helpers.go
Comment (aneeshnema) [reply mesheryctl/pkg/utils/helpers.go:724]: If we are planning to add temporary-context to other commands, then I think you are correct. Currently, all the other commands except `resetMesheryConfig` download docker-compose file only if it is not found. Ideally, it should download the file if it does not exist or **if the docker-compose file does not match the current-context** (especially needed for temporary-context). The second part is not handled at the moment. Though making mesheryctl always download docker-compose file will be a pain for users not connected to the internet.

Comment (navendu-pottekkat) [reply mesheryctl/pkg/utils/helpers.go:724]: The `--context` flag would be ideally a global flag that is available in all of the mesheryctl commands. This has yet to be implemented but it will be down the road. So we should be adding it in such a way that we won't have to rewrite it in the future.

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [inline mesheryctl/pkg/utils/helpers.go:726]: Should the version also be "edge"? Or should it be "latest"? The channel and version being "edge" feels a bit off. What does everyone suggest?

**Review by navendu-pottekkat — APPROVED** [review]
This looks good. We can use the `GetCurrentContext` and the `DownloadDockerComposeFile` throughout the other commands.

Screenshot from 2021-02-12 09-00-39

One small change I would suggest here is to replace the second _"dev" context does not exist_ with just "context does not exist" since we are already mentioning about the "dev" context in the line above.

Adding comments on the `reset.go` file and a comment on using the `force` bool in the function in `helpers.go` would be helpful in the future.

Rest LGTM and works as expected when testing. Good work!

**Comment by navendu-pottekkat** [comment]
@anirudhjain75 Could you also please check this out before we merge?

**Comment by anirudhjain75** [comment]
Sure. Give me sometime

File: mesheryctl/pkg/utils/helpers.go
Comment (anirudhjain75) [reply mesheryctl/pkg/utils/helpers.go:726]: Yea this might raise an issue.. Currently the versions are edge-latest and stable-latest. so the version should either be set as latest or be leftout. first alternative is better.

Comment (anirudhjain75) [reply mesheryctl/pkg/utils/helpers.go:726]: Also you dont need a GetCurrentContext function... The mesheryctl config class already has that inplace. A good place to look into is config.go

**Review by anirudhjain75 — CHANGES_REQUESTED** [review]
Hey @aneeshnema great stuff here. Just have a few things you might want to address :).

File: mesheryctl/pkg/utils/helpers.go
Comment (anirudhjain75) [reply mesheryctl/pkg/utils/helpers.go:726]: if a change is required, feel free to edit config.go instead

**Comment by aneeshnema** [comment]
@leecalcote some of your commits are not signed-off, hence my DCO has failed. The instructions on the DCO page will signs-off with my identity but the author remains the original person. Thus it will fail again due to mismatch.

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [inline mesheryctl/pkg/utils/helpers.go:699]:
```suggestion
	// if the user mentions a temporary context with the -c flag, change the current-context to temporary context
```

**Review by navendu-pottekkat — APPROVED** [review]
Looks good. Thanks for making the changes suggested. @leecalcote  @anirudhjain75  I think we can merge this.

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [inline mesheryctl/pkg/utils/helpers.go:716]:
```suggestion
	return "", config.Context{}, errors.Errorf("\n context does not exist")
```

File: mesheryctl/pkg/utils/helpers.go
Comment (aneeshnema) [reply mesheryctl/pkg/utils/helpers.go]: check out the updated code, I don't think we should get rid of GetCurrentContext. If we remove it, we will need to add duplicate code at many places

**Comment by leecalcote** [comment]
@aneeshnema it looks like I'm holding you back. 😉 It's frustrating that commits performed through GitHub's web interface don't include sign-offs despite the user being authenticated. Fixing DCO...

File: mesheryctl/pkg/utils/helpers.go
Comment (navendu-pottekkat) [reply mesheryctl/pkg/utils/helpers.go]: @aneeshnema Yes, that would be best

**Review by navendu-pottekkat — APPROVED** [review]
@anirudhjain75 @leecalcote LGTM. We can merge the changes and let @aneeshnema carry on with the awesome work he has been doing.

**Comment by anirudhjain75** [comment]
Merging conflict I guess

**Comment by aneeshnema** [comment]
resolved the conflict ✔ @anirudhjain75 
but now I have DCO problems again. Can someone help? @leecalcote

**Comment by anirudhjain75** [comment]
You can simple foolow steps from dco

**Comment by aneeshnema** [comment]
> @leecalcote some of your commits are not signed-off, hence my DCO has failed. The instructions on the DCO page will signs-off with my identity but the author remains the original person. Thus it will fail again due to mismatch.

@anirudhjain75

**Comment by leecalcote** [comment]
@aneeshnema the likely culprit here is a branch being out of date. You’ll want to sync your fork and branch with the latest upstream. Ideally, those commits are not included in your PR. This section will help - https://github.com/layer5io/meshery/blob/master/CONTRIBUTING-gitflow.md#verify-the-new-

**Comment by anirudhjain75** [comment]
@aneeshnema are the issues with end to end test on ci because of your changes ?

**Comment by aneeshnema** [comment]
@anirudhjain75 nope, I think it's a time out error
https://github.com/layer5io/meshery/pull/2397/checks?check_run_id=1890292464
> upstream request timeoutupstream connect error or disconnect/reset before headers. reset reason: local resetupstream connect error or disconnect/reset before headers. reset reason: local resetupstream connect error or disconnect/reset before headers. reset reason: local resetupstream connect error or disconnect/reset before headers. reset reason: connection failureupstream connect error or disconnect/reset before headers. reset reason: local reset

**Comment by anirudhjain75** [comment]
@leecalcote can you confirm this if not from this PR.

**Comment by leecalcote** [comment]
You're right, gents. This failure isn't do to the code in this PR.

**Pull request merged**
